webhooks/cluster.go

```diff
@@ -182,34 +191,33 @@ func (webhook *Cluster) validateTopology(ctx context.Context, old, new *clusterv
 		)
 	}
 
-	// MachineDeployment names must be unique.
-	if new.Spec.Topology.Workers != nil {
-		names := sets.String{}
-		for _, md := range new.Spec.Topology.Workers.MachineDeployments {
-			if names.Has(md.Name) {
-				allErrs = append(allErrs,
-					field.Invalid(
-						field.NewPath("spec", "topology", "workers", "machineDeployments"),
-						md,
-						fmt.Sprintf("MachineDeployment names should be unique. MachineDeployment with name %q is defined more than once.", md.Name),
-					),
-				)
-			}
-			names.Insert(md.Name)
-		}
+	// clusterClass must exist.
+	clusterClass := &clusterv1.ClusterClass{}
+	// Check to see if the ClusterClass referenced in the Cluster currently exists.
+	if err := webhook.Client.Get(ctx, client.ObjectKey{Namespace: new.Namespace, Name: new.Spec.Topology.Class}, clusterClass); err != nil {
+		allErrs = append(
+			allErrs, field.Invalid(
+				field.NewPath("spec", "topology", "class"),
+				new.Name,
+				"ClusterClass could not be found"))
+		return allErrs
 	}
 
+	allErrs = append(allErrs, check.MachineDeploymentClassesAreUnique(clusterClass)...)
+	allErrs = append(allErrs, check.MachineDeploymentTopologiesAreUniqueAndDefinedInClusterClass(new, clusterClass)...)
+
 	if old != nil { // On update
-		// Class could not be mutated.
-		if new.Spec.Topology.Class != old.Spec.Topology.Class {
+		// Topology or Class can not be added on update.
+		if old.Spec.Topology == nil || len(old.Spec.Topology.Class) == 0 {
```

Comment:
What about changing this check so the users can set Spec.Topology on a existing cluster (not only the class field)

Reply:
Is this the unmanaged to managed scenario? I'm blocking that on purpose here as I'm not sure how we want to deal with it. I thought we'd want to take a look at that change down the line rather than enabling it here.

Reply:
@fabriziopandini  Not sure if I understood, do you want to allow enabling ClusterClass on a previously unmanaged Cluster?

Reply:
I think that we should:
1. avoid to pass from a cluster having a topology to a cluster without it and vice versa.
2. when there is topology, we should ensure class has a value and it doesn't get blanked out.

The current implementation seems to cover only 2

Reply:
Here we're checking for the unmanaged to managed case i.e. new.Topology is defined but old.Topology isn't defined. The managed to unmanaged case is being checked on L141

Reply:
if it is already covered, then I'm ok.
thanks for clarifying